Add Playwright tests for All Reviews, Reviews by Product and Reviews by Category blocks

Aljullu opened the pull request:
<!-- Please do not remove any information from this pull request. Instead, add N/A or leave blank if not applicable -->

## What

~Blocked by https://github.com/woocommerce/woocommerce-blocks/pull/11976.~

Fixes https://github.com/woocommerce/woocommerce/issues/42065.

This PR adds Playwright tests to the All Reviews, Reviews by Product and Reviews by Category blocks and removes the Puppeteer tests for the same blocks.

## Why

Hopefully, these tests would have caught these issues: https://github.com/woocommerce/woocommerce-blocks/issues/11912, https://github.com/woocommerce/woocommerce-blocks/issues/11917 and https://github.com/woocommerce/woocommerce-blocks/issues/11974.

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. Verify Playwright e2e tests pass.

* [x] Do not include in the Testing Notes
* [ ] Should be tested by the development team exclusively

## Screenshots or screencast

| Before | After |
| ------ | ----- |
|        |       |

## WooCommerce Visibility

Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [x] N/A

## Checklist

Required:

* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:

* [ ] This PR has a UI change and has been cross-browser tested at different viewport sizes on both the frontend and in the editor.
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog
<!-- Provide a brief, descriptive summary of the changes in this PR. Include potential impacts on different parts of the product. Example: "Updated the checkout process to streamline the experience for users and reduce the number of steps." -->

> Add suggested changelog entry here.

Aljullu committed:
Add All Reviews block tests

Aljullu committed:
Move publishAndVisitPost() to utils file

Aljullu committed:
Typos

Aljullu committed:
Add Reviews by Product block tests

Aljullu committed:
Add Reviews by Category block tests

Aljullu committed:
Remove Puppeteer tests for reviews blocks

Aljullu commented on tests/e2e/global-setup.ts:
Changes in this file are based on the same file in WooCommerce core:

https://github.com/woocommerce/woocommerce/blob/759d903d925df7ec00e1693d179444ceb5112e8c/plugins/woocommerce/tests/e2e-pw/global-setup.js

Aljullu committed:
Fix Customer Account test

imanish003 commented on tests/e2e/tests/all-reviews/all-reviews.block_theme.spec.ts:
```diff
@@ -0,0 +1,120 @@
+
+	// Create product and reviews.
+	test.beforeAll( async ( { baseURL } ) => {
+		const api = new WooCommerceRestApi( {
+			url: baseURL,
+			consumerKey: process.env.CONSUMER_KEY,
+			consumerSecret: process.env.CONSUMER_SECRET,
+			version: 'wc/v3',
+		} );
```
Interesting! I've got a couple of questions about using "WooCommerceRestApi" in our Playwright tests:

1. Is it okay to integrate "WooCommerceRestApi" for testing purposes? Specifically, we have some tests for the Product Collection block that require data not available in the sample. Is it a good idea to use this API to create new products before the tests run and then clean them up with an `afterAll`? For instance, we need to test scenarios like “Out of Stock” products, as you can see [here](https://github.com/woocommerce/woocommerce-blocks/blob/9f76eb95966dce5b70c76e8819e7ee23cec56d13/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts#L206)
2. Also, do you think there will be any impact when we run all tests in parallel? Currently, our tests run serially, but I'm guessing we'll be shifting to parallel testing soon. What are your thoughts on this?

Aljullu replied:
Good questions, I'm not 100% sure, to be honest. I copied this approach [from WC core](https://github.com/woocommerce/woocommerce/blob/f35d52203ea49443113a2609f787981ba2958e5f/plugins/woocommerce/tests/e2e-pw/tests/shopper/cart.spec.js#L9-L49), but I'm not sure whether there is a better way of doing that. Pinging @dinhtungdu as he knows a lot about e2e tests.

imanish003 replied:
Understood. I raised this question primarily to explore whether we could apply the same approach to other areas of our codebase, including the tests for Product Collection. Gaining some clarity on this aspect would be extremely helpful.

@sunyatasattva I would really appreciate your input here 🙂

dinhtungdu replied:
We're still searching for a good way to prepare the data for testing. I think using WooCommerce REST API is a good approach we should consider standardizing. For now we're creating test data at the provisioning time, and there is clear needs to create test data/fixture at the test run time. Manish also did some research here for the Seville meetup: per0F9-13G-p2.

> Currently, our tests run serially, but I'm guessing we'll be shifting to parallel testing soon.

Even we aim to run the test in parallel and we should write tests in isolation, I think we should only enable parallelism on local, not CI. According to the Playwright docs:

> We recommend setting [workers](https://playwright.dev/docs/api/class-testconfig#test-config-workers) to "1" in CI environments to prioritize stability and reproducibility. Running tests sequentially ensures each test gets the full system resources, avoiding potential conflicts. However, if you have a powerful self-hosted CI system, you may enable [parallel](https://playwright.dev/docs/test-parallel) tests. For wider parallelization, consider [sharding](https://playwright.dev/docs/test-parallel#shard-tests-between-multiple-machines) - distributing tests across multiple CI jobs.
> https://playwright.dev/docs/ci#workers

sunyatasattva replied:
From my personal point of view, we should try to find an alternative approach to using the REST API on `setup` and `teardown` phases of the test. We have all found that E2E tests already take a long time, and creating fixtures on a test-by-test basis using the API is going to increase the run time of the test. Also, using the API is going to create another dependency, and thus another point of failure (i.e. increase flakiness).

With that said, standardization of course is of great importance. So, establishing repeatable patterns should also be a design priority.

In this case: is there a reason why can't we add a product with reviews on our initial fixture to test these use-cases?

imanish003 replied:
> In this case: is there a reason why can't we add a product with reviews on our initial fixture to test these use-cases?

AFAIK there was a related discussion #10099, where I proposed a standard approach for adding new fixture data, leading to task #10189. The main reason we haven't done this yet is our limited bandwidth during the cooldown period. Maybe we can prioritize this task to work on it during the cooldown 🤔

Aljullu replied:
Thanks for all your feedback, folks! :raised_hands: 

> In this case: is there a reason why can't we add a product with reviews on our initial fixture to test these use-cases?

Not really, it was mostly because I took WC core tests as an example. I guess one advantage of using the API is that we can have different tests using different test data. Ie: if at some point we wanted to test how the Reviews blocks behave when there are no reviews in the store, it would be easier to do so if there are no reviews created during the setup.

In any case, that's not the case right now, as we are only testing these blocks when _there are_ reviews. :slightly_smiling_face: So in cc89fa6f71a004052e37535ba1ea993c78b87bb7 I refactored this PR to create the reviews on setup instead of at the beginning of the test. I also removed `WCRestApiUtils` for now, as it's no longer necessary.

@imanish003 I will need to ask for another review. :pray:

imanish003 commented on tests/e2e/tests/all-reviews/all-reviews.block_theme.spec.ts:
```diff
@@ -0,0 +1,120 @@
+		await admin.createNewPost();
+		await editor.insertBlock( { name: blockData.name } );
+
+		await expect( page.getByText( 'Nice album!' ) ).toBeVisible();
```
Nit: Should we think about creating constants for strings like 'Nice album!'? What do you think?

Aljullu replied:
Done in d769357e58fbe9d6dd7400cf5c2b098c44d02edc.

imanish003 commented on tests/e2e/utils/editor/editor-utils.page.ts:
```diff
@@ -320,4 +320,11 @@ export class EditorUtils {
 			.waitFor();
 	}
+
+	async publishAndVisitPost() {
+		await this.editor.publishPost();
+		const url = new URL( this.page.url() );
+		const postId = url.searchParams.get( 'post' );
+		await this.page.goto( `/?p=${ postId }`, { waitUntil: 'commit' } );
+	}
```
Thanks for making this function reusable 🚀

imanish003 commented on tests/e2e/tests/product-collection/compatibility-layer.block_theme.side_effects.spec.ts:
I appreciate you taking the time to fix the spelling errors. Thank you!

imanish003 commented on tests/e2e/tests/reviews-by-category/reviews-by-category.block_theme.spec.ts:
Nit: I've noticed a few TypeScript errors. What do you think about either fixing these or adding an @expect-error comment?

Aljullu replied:
Good catch. I added types to the `reponse` from API calls in cf6d39bf67011f8767914080f5687a35da3e806d. There is still the issue that `@woocommerce/woocommerce-rest-api` is not typed. We have the same issue in [other parts of the codebase](https://github.com/search?q=repo%3Awoocommerce%2Fwoocommerce-blocks%20%40woocommerce%2Fwoocommerce-rest-api&type=code), but I'm not sure how to solve it easily. :thinking:

imanish003 replied:
I took a quick look at “Definitely Typed” repo. Looks like we can use `@types/woocommerce__woocommerce-rest-api` from there as you can see [here](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/e7b275ffdb647d8efec72ce0f240fd763eb49a02/types/woocommerce__woocommerce-rest-api). What do you think?

Aljullu replied:
Oh, you are right. I installed it in 0c0bf3fe09b1db7d68322e5a667e192220c87e7c. I also refactored the code a bit in fcc02977f55e72a74f2a7d6f965dbfda2799a862 to create a util for the communication with the API, so we don't need to do that in every test file.

imanish003 replied:
I really appreciate the reusable utility `WCRestApiUtils` that you've developed. It's a great addition! 🙌🏻

imanish003 reviewed (approved):
I've successfully executed all the Playwright tests on my local machine, and they are passing. A big thank you for transitioning the tests to Playwright! 🚀

I've left a few minor comments, but I'm pre-approving the PR since they aren't blockers.

Aljullu committed:
Use constants instead of hardcoded strings for reviews content

Aljullu committed:
Add type to reponse argument

Aljullu commented:
Thanks for the review, @imanish003, I implemented some of the changes. Would you mind taking another look?

imanish003 reviewed (approved):
Thanks for quickly making the changes. I've reviewed the changes and everything seems great 🚀🙌🏻. I've also responded to your comments 🤝.

Sidenote: Looks like PR is missing Sprint, Estimate etc. 🙂

Aljullu committed:
Install WC rest API types and refactor API creation in tests

Aljullu committed:
Create WCRestApiUtils

Aljullu commented:
Thanks for the reviews, @imanish003! I added some more changes so I asked for another look. :pray: 

> Sidenote: Looks like PR is missing Sprint, Estimate etc. 🙂

Added. :+1:

imanish003 reviewed (approved):
Hi @Aljullu,

Changes looks good to me. Also, I really appreciate the reusable utility WCRestApiUtils that you've developed. It's a great addition! 🙌🏻

P.S.: There's an ongoing discussion in the https://github.com/woocommerce/woocommerce-blocks/pull/11975#discussion_r1410470663. I don’t see it as a blocker, so I'm approving the PR.

Aljullu committed:
Create product reviews on setup instead of using the WC Rest API

Aljullu commented:
This PR has been ported to the monorepo: https://github.com/woocommerce/woocommerce-blocks/pull/11975.
